fiftyone/utils/eval/detection.py

```diff
@@ -193,34 +243,33 @@ def evaluate_detections(
     else:
         _samples = samples.select_fields([gt_field, pred_field])
 
+    def _map_fnc(sample) -> List[Any]:
+        return _compute_matches_single(
+            sample,
+            eval_method,
+            eval_key,
+            save=save,
+            processing_frames=processing_frames,
+            tp_field=tp_field,
+            fp_field=fp_field,
+            fn_field=fn_field,
+        )
+
     matches = []
-    logger.info("Evaluating detections...")
-    for sample in _samples.iter_samples(progress=progress, autosave=save):
-        if processing_frames:
-            docs = sample.frames.values()
-        else:
-            docs = [sample]
-
-        sample_tp = 0
-        sample_fp = 0
-        sample_fn = 0
-        for doc in docs:
-            doc_matches = eval_method.evaluate(doc, eval_key=eval_key)
-            matches.extend(doc_matches)
-            tp, fp, fn = _tally_matches(doc_matches)
-            sample_tp += tp
-            sample_fp += fp
-            sample_fn += fn
-
-            if processing_frames and save:
-                doc[tp_field] = tp
-                doc[fp_field] = fp
-                doc[fn_field] = fn
-
-        if save:
-            sample[tp_field] = sample_tp
-            sample[fp_field] = sample_fp
-            sample[fn_field] = sample_fn
+    if parallelize_method or workers:
+        for _, result in _samples.map_samples(
+            _map_fnc,
+            workers=workers,
+            batch_method=batch_method,
+            progress=progress,
+            parallelize_method=parallelize_method,
+            save=save,
+        ):
+            matches.extend(result)
+    else:
+        for sample in _samples.iter_samples(progress=progress, autosave=save):
```

Comment:
Since map_samples has a fallback for sequential processing, can this work correctly if we always call map_samples, and avoid having this additional “else” with iter_samples?

Reply:
Good idea - I was thinking about it too! In this way, we can remove all instances of `iter_samples` from the code base unless it is something `map_samples` can't solve it (applying_model for example which requires access to an in-memory model and dataloader).